x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -220,379 +198,410 @@ export function App({
-                );
-                setState((s) => ({
-                  ...s,
-                  isLoading: false,
-                  persistedDoc: doc,
-                  lastKnownDoc: doc,
-                  query: doc.state.query,
-                  indexPatternsForTopNav: indexPatterns,
-                }));
-              })
-              .catch((e) => {
-                setState((s) => ({ ...s, isLoading: false }));
+  useEffect(() => {
+    if (
+      !initialInput ||
+      (attributeService.inputIsRefType(initialInput) &&
+        initialInput.savedObjectId === state.persistedDoc?.savedObjectId)
+    ) {
+      return;
+    }
 
-                redirectTo();
-              });
+    setState((s) => ({ ...s, isLoading: true }));
+    attributeService
+      .unwrapAttributes(initialInput)
+      .then((attributes) => {
+        if (!initialInput) {
+          return;
+        }
+        const doc = {
+          ...initialInput,
+          ...attributes,
+          type: LENS_EMBEDDABLE_TYPE,
+        };
+
+        if (attributeService.inputIsRefType(initialInput)) {
+          chrome.recentlyAccessed.add(
+            getFullPath(initialInput.savedObjectId),
+            attributes.title,
+            initialInput.savedObjectId
+          );
+        }
+        getAllIndexPatterns(
+          _.uniq(doc.references.filter(({ type }) => type === 'index-pattern').map(({ id }) => id)),
+          data.indexPatterns,
+          notifications
+        )
+          .then((indexPatterns) => {
+            // Don't overwrite any pinned filters
+            data.query.filterManager.setAppFilters(
+              injectFilterReferences(doc.state.filters, doc.references)
+            );
+            setState((s) => ({
+              ...s,
+              isLoading: false,
+              persistedDoc: doc,
+              lastKnownDoc: doc,
+              query: doc.state.query,
+              indexPatternsForTopNav: indexPatterns,
+            }));
           })
           .catch((e) => {
             setState((s) => ({ ...s, isLoading: false }));
-
-            core.notifications.toasts.addDanger(
-              i18n.translate('xpack.lens.app.docLoadingError', {
-                defaultMessage: 'Error loading saved document',
-              })
-            );
-
             redirectTo();
           });
-      }
-    },
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [
-      core.notifications,
-      data.indexPatterns,
-      data.query.filterManager,
-      docId,
-      // TODO: These dependencies are changing too often
-      // docStorage,
-      // redirectTo,
-      // state.persistedDoc,
-    ]
-  );
+      })
+      .catch((e) => {
+        setState((s) => ({ ...s, isLoading: false }));
+        notifications.toasts.addDanger(
+          i18n.translate('xpack.lens.app.docLoadingError', {
+            defaultMessage: 'Error loading saved document',
+          })
+        );
+
+        redirectTo();
+      });
```

Comment:
This `.catch` function looks like a copy+paste of the one on line 297, why?

Reply:
There are two catch cases here. The outer one is to catch a document unwrap failure, the inner one catches a failure on `getAllIndexPatterns`. It is the same on master, and I haven't thought of a good way to consolidate them.